client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -202,21 +201,34 @@ private boolean checkPushBlacklisted(
   private void submitRetryPushData(
       String applicationId,
       int shuffleId,
-      int mapId,
-      int attemptId,
       byte[] body,
       int batchId,
-      PartitionLocation loc,
       RpcResponseCallback wrappedCallback,
       PushState pushState,
-      StatusCode cause,
+      ReviveRequest request,
       int remainReviveTimes) {
+    int mapId = request.mapId;
+    int attemptId = request.attemptId;
+    PartitionLocation loc = request.loc;
+    StatusCode cause = request.cause;
     int partitionId = loc.getId();
-    if (!revive(
-        applicationId, shuffleId, mapId, attemptId, partitionId, loc.getEpoch(), loc, cause)) {
-      wrappedCallback.onFailure(
-          new CelebornIOException(cause + " then revive but " + StatusCode.REVIVE_FAILED));
-    } else if (mapperEnded(shuffleId, mapId)) {
+    long reviveWaitTime =
+        conf.clientRpcRequestPartitionLocationRpcAskTimeout().duration().toMillis() + 1000;
+    final long delta = 50;
+    long accumulatedTime = 0;
+    while (request.reviveStatus == StatusCode.REVIVE_INITIALIZED
+        && accumulatedTime <= reviveWaitTime) {
+      try {
+        Thread.sleep(50);
+        accumulatedTime += delta;
+      } catch (InterruptedException e) {
+        logger.error("Interrupted while waiting for Revive result!");
+        wrappedCallback.onFailure(
+            new CelebornIOException(cause + " then revive but " + StatusCode.REVIVE_FAILED));
```

Comment:
Should not return this here? Refer to https://github.com/apache/incubator-celeborn/pull/1486